### PR TITLE
update 'prepview' for k8 v1.24 & fetch PR-495

### PR DESCRIPTION
**Pull request by satya-dillikar**
- CLUSTER_TOKEN value (kubectl command) is changed in Kubernetes cluster v1.24 or later.

> More details [here](https://docs.vmware.com/en/VMware-Tanzu-Application-Platform/1.4/tap/tap-gui-cluster-view-setup.html).
> This was causing prepview to fail internally. The fix updates the new token command/resources.

- To fetch changes in [PR-495](https://github.com/aws-quickstart/quickstart-amazon-eks/pull/495), updated the quickstart-amazon-eks submodule to commit#17cddfb.
These have fixes to cleanup all resources on the CFT stack delete operation
> 
> This PR has breaking changes, params have changed.
> 1) ALBIngressController to LoadBalancerController
> 2) MonitoringStack is removed.

### CODE REVIEW

**Comment by satya-dillikar** [comment]
@hoegaarden , please check why Concourse CI build failed.
Thanks

**Comment by hoegaarden** [comment]
You can check that yourself by following the link here:
image

File: tap-setup-scripts/src/functions.sh
Comment (hoegaarden) [inline tap-setup-scripts/src/functions.sh:498]: *nit*: I am a bit concerned about those "random sleeps" in our code. But I guess it's the easiest way to get unblocked.

File: tap-setup-scripts/src/functions.sh
Comment (hoegaarden) [inline tap-setup-scripts/src/functions.sh:702]: *hint*: This could also be done with `kubectl` only, with a crisp `kubectl -n tap-gui get secret tap-gui-viewer -o go-template='{{ .data.token | base64decode }}'`

File: tap-setup-scripts/src/tap-main.sh
Comment (hoegaarden) [inline tap-setup-scripts/src/tap-main.sh:40]: *question*: Not sure why we do a `kubectl create ... || true` here? My assumption is, we do it to not fail if the secret already exists. But wouldn't a `kubdectl apply ...` be better and not swallow potential issues/errors?

File: tap-setup-scripts/src/resources/tap-gui-viewer-secret.yaml
Comment (hoegaarden) [inline tap-setup-scripts/src/resources/tap-gui-viewer-secret.yaml:3]: *question*: Instead of creating that secret here, could we add this as an overlay to the `tap-gui` package?

With that, we wouldn't have to do those "one-off" `kubectl` calls to create and delete the secret; the kapp-controller should handle all that for us, and we'd don't have to "reimplement" all that in our cloud init script.

Most/all(?) parts for that should already be there and what I am thinking of is something like:
1. add another overlay secret to [package_overlay_secrets](https://github.com/vmware-tap-on-public-cloud/quickstart-vmware-tanzu-application-platform/blob/d148889f3637bced3180a3a31065e7d2248edd3f/tap-setup-scripts/src/resources/package_overlay_secrets.yaml), similar to the the [telemetry overlays](https://github.com/vmware-tap-on-public-cloud/quickstart-vmware-tanzu-application-platform/blob/d148889f3637bced3180a3a31065e7d2248edd3f/tap-setup-scripts/src/resources/package_overlay_secrets.yaml#L24-L32)
2. In [package_overlay_apply](https://github.com/vmware-tap-on-public-cloud/quickstart-vmware-tanzu-application-platform/blob/d148889f3637bced3180a3a31065e7d2248edd3f/tap-setup-scripts/src/resources/package_overlay_apply.yaml) add that overlay secret to the `pkgi`, similar to the [telemetry overlays](https://github.com/vmware-tap-on-public-cloud/quickstart-vmware-tanzu-application-platform/blob/d148889f3637bced3180a3a31065e7d2248edd3f/tap-setup-scripts/src/resources/package_overlay_apply.yaml#L8-L12)

File: tap-setup-scripts/src/tap-main.sh
Comment (hoegaarden) [reply tap-setup-scripts/src/tap-main.sh:40]: These would go away, I guess, if we'd add the token secret as an overlay to `tap-gui`, see other comment.

File: tap-setup-scripts/src/functions.sh
Comment (hoegaarden) [reply tap-setup-scripts/src/functions.sh:498]: *question*: Why are we waiting here anyway? What are we waiting for?

Are we trying to wait for another reconciliation cycle of the `tap` `pkgi`, so that we can then assert on the state of the packages?

Wouldn't it be more useful to e.g.
- "Kick" the `tap` `pkgi`/`app`
    - sidenote: "kick" in the sense of adding / updating a label or annotation on the resource; I wouldn't want to introduce a dependency on the `kctrl` tool, which would have a literal `kick` subcommand.
- `kubectl wait` for its status / some jsonpath to change

instead of "randomly" waiting for 1 minute?

**Review by hoegaarden — CHANGES_REQUESTED** [review]
1. Can we, in future, please not shoehorn multiple unrelated things (in this case: bumping the EKS stack & fixing the token handling; they are completely unrelated & independnt, no?) into one PR but rather have them separately?
2. TBH, I am not sure things seem to have worked "sometimes", even without this token handling improvement. However, I didn't go back and check previous deploys in CI on that, this needs to be fixed regardless for 1.24 clusters.
3. I think we should really consider relying more on overlays instead of making the cloud init more complex

File: tap-setup-scripts/src/functions.sh
Comment (satya-dillikar) [reply tap-setup-scripts/src/functions.sh:498]: Ideally, this wait is not required. I added it to make sure that the other reconciliation cycle is settled before we move to check again. Might be helpful in some rare-condition.

File: tap-setup-scripts/src/functions.sh
Comment (satya-dillikar) [reply tap-setup-scripts/src/functions.sh:702]: Thanks for the info. I guess we have used jq mostly.  Prefer not to change now.

File: tap-setup-scripts/src/tap-main.sh
Comment (satya-dillikar) [reply tap-setup-scripts/src/tap-main.sh:40]: Yes, we can use "apply" instead of "create". I  have not seen any significant differences in the behaviors for all our yaml resources till now.

File: tap-setup-scripts/src/resources/tap-gui-viewer-secret.yaml
Comment (satya-dillikar) [reply tap-setup-scripts/src/resources/tap-gui-viewer-secret.yaml:3]: yes, there are multiple ways to do the same thing. An overlay is one of them.  I guess we can do that in another exercise in a different PR.

**Comment by satya-dillikar** [comment]
> 1. Can we, in future, please not shoehorn multiple unrelated things (in this case: bumping the EKS stack & fixing the token handling; they are completely unrelated & independnt, no?) into one PR but rather have them separately?
> 2. TBH, I am not sure things seem to have worked "sometimes", even without this token handling improvement. However, I didn't go back and check previous deploys in CI on that, this needs to be fixed regardless for 1.24 clusters.
> 3. I think we should really consider relying more on overlays instead of making the cloud init more complex

- This PR is open and waiting for approval for many days. Once I did my second 'push'   from the same local folder, GitHub automatically added the commits to this open PR. Hence you see all diffs here.
- is overlays complex or cloud-init complex: I think this is just the perception and related to personal familiarity  & use.  Whatever we pick, I think we should be consistent. This reminds me to add a coding guidelines doc for this repo.

File: tap-setup-scripts/src/functions.sh
Comment (hoegaarden) [reply tap-setup-scripts/src/functions.sh:702]: *non-blocking*

If feasible, I think no pipeline should still be preferred as I don't think we run with `pipefail` nor do we run `jq` with `-e`, and thus just swallows a lot of error cases.

File: tap-setup-scripts/src/tap-main.sh
Comment (hoegaarden) [reply tap-setup-scripts/src/tap-main.sh:40]: I am always uneasy when I see something like `|| true` without explaination / good reason / ... because they just swallow all errors, also the ones I potentially don't want to ginore.

File: tap-setup-scripts/src/functions.sh
Comment (hoegaarden) [reply tap-setup-scripts/src/functions.sh:498]: > [...] reconciliation cycle is settled [...]

That's why I proposed to kick the app and `kubectl wait` for the reconciliation to finish.

Going forward (but not blocking for this PR), I think we should try hard to avoid sleeps if possible.

File: tap-setup-scripts/src/resources/tap-gui-viewer-secret.yaml
Comment (hoegaarden) [reply tap-setup-scripts/src/resources/tap-gui-viewer-secret.yaml:3]: *non-blocking*

To add more context why I believe overlays should generally be preferred:

- any change we do, any objects we modify, any objects we add, are now part of the `pkgi` and its `app`. Thus the kapp-controller handles their lifecycle, e.g. cleanup.
- with overlays we can describe the desired state from the get go, thus we don't have to deploy something with a "not desired" state, to later change things up, redeploy, and hope it can transition from that intermediate state to the actual desired state.
- with the use of kapp's changegroups/changerules we can express ordering of upserts & deletes of objects. With kapp's waitrules we can have influence on when we deem objects to have reconciled correctly. We can "just" instruct the kapp-controller to do these kinds of things for us, we don't have to implement them with busy loops.

While the later points are not relevant for this specific case, the first one is.

File: ci/qs-test/taskcat_overlay.yml
Comment (hoegaarden) [inline ci/qs-test/taskcat_overlay.yml:46]: As discussed on slack: This is only part of the change we got to do. This change will break other parts of the pipeline.

**Review by hoegaarden — APPROVED** [review]
> This PR is open and waiting for approval for many days. Once I did my second 'push' from the same local folder, GitHub automatically added the commits to this open PR. Hence you see all diffs here.

Yeah, that's how github works: A PR is based on a branch. Any commits you push to that branch will be added to that PR. Thus, if you have multiple Independent things, you got to separate them in different branches and create 2 distinct PRs from those different branches.

> is overlays complex or cloud-init complex: I think this is just the perception and related to personal familiarity & use. Whatever we pick, I think we should be consistent.

Totally agree, that complexity (in parts) is subjective. I also prefer to be consistent, when feasible. But not when it's blocking improvements. Using better suited, tailored tools (e.g. kapp-controller vs. bash) is an improvement in my book. Any line we can move from the imperative cloud init to e.g. declarative pkgi/app, too.

Maybe it's not clear what my north star is:
Ideally, I want the cloud init to "just" prepare the desired state and then push that to the system; and have the system sort itself out. This would make our thing smaller, easier to reason about and test. Will we ever completely get there? Maybe not. But there are quite some opportunities where we can do exactly that.


Having said all that, and despite the issue with the taskcat overlay change breaking other parts, I will merge this PR anyway and deal with the fallout later. I want us to have the new version of the EKS stack everywhere, so that we get out of this incompatibilities with the (account) shared resources.

**Comment by hoegaarden** [comment]
Merging now, with me as an explicit approver, and Satya, as the PR author, as an implicit approver ; accepting the failure of the `publish-bucket` job, which I've paused for now.

File: ci/qs-test/taskcat_overlay.yml
Comment (hoegaarden) [reply ci/qs-test/taskcat_overlay.yml:46]: To be clear about what I mean: [see here](https://runway-ci-sfo.eng.vmware.com/teams/tappc/pipelines/qs-test/jobs/prepare-bucket/builds/51#L63f6a68f:1911:1922)
In this instance it doesn't matter, it's just printing garbage, however in the `publish-bucket` job, it would fail, because there is no region `$[taskcat_current_region]`

**Pull request merged**
